### ISSUE TITLE
feat: React 18 support

### ISSUE THREAD

**Issue by jorroll** [solidjs/react-solid-state#4]
It appears that `react-solid-state` doesn't support the new Renderer in React 18.

Using the old render function with React 18 still works, but switching to the [new render function in React 18 ](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis)doesn't work.

- [Working React 17 example](https://stackblitz.com/edit/react-ts-seujby?file=index.tsx)
- [Broken React 18 example](https://stackblitz.com/edit/react-ts-j9gr3b?file=index.tsx).

**Comment by ryansolid** [comment]
Yeah I suspected this would happen. I just haven't been up on the new primitives. React 18 basically requires a new implementation for effectful state libraries.

**Comment by jorroll** [comment]
Here's a rough POC fix
- [Working React 18 example](https://stackblitz.com/edit/react-ts-qggoxr?file=react-solid-state.ts)

I suspect React 18 currently isn't working because `setState()` calls are [now being batched](https://reactjs.org/blog/2022/03/29/react-v18.html#new-feature-automatic-batching). In my simple proof-of-concept fix, I'm using the new [`useSyncExternalStore` hook](https://reactjs.org/blog/2022/03/29/react-v18.html#usesyncexternalstore) to force updates instead of a call to `setState()`.

```ts
function useNewForceUpdate() {
  const [tick, setTick] = rState(0);
  const fn = rCallback(() => {
    setTick((t) => t + 1);
    return () => {};
  }, []);
  const fn2 = rCallback(() => tick, []);
  const state = rSyncExternalStore(fn, fn2);
  return fn;
}
```

I expect that this can be cleaned up, but I was mostly concerned with figuring out what was going on.

**Comment by ryansolid** [comment]
Awesome. Thanks for looking into this. I guess MobX would have a similar mechanism. If this works tempted to do this. I suppose couldn't hurt to see where they landed, now that we know what to look for.

**Comment by jorroll** [comment]
I took a look at MobX's current React 18 support. While MobX currently works with React 18, it doesn't yet play nicely with suspense/concurrent features ([see comments on the PR introducing React 18 support](https://github.com/mobxjs/mobx/pull/3387)). The author of that PR calls out using `useSyncExternalStore` as their most likely path forward to greater React 18 support, but it sounds like they'll need to do a larger refactor to make that work and they want to wait and get more info about real world usage first. [Discussion is ongoing here](https://github.com/mobxjs/mobx/issues/2526).

I'll note that it appears that MobX already worked with React 18's change detection, as the PR which "adds" support for React 18 only updates the `package.json` to list React 18 support and updates a bunch of internal tests. None of the distributed MobX package code needed to be updated though (well ok, there was one small typescript typing change, but it didn't affect the javascript). So currently MobX doesn't use `useSyncExternalStore` or any of the new React 18 hooks (which has allowed them to add React 18 support in a backwards-compatible minor release).

All of this is to say that, at the moment, it doesn't look like MobX can provide much insight but [we should keep an eye on this issue](https://github.com/mobxjs/mobx/issues/2526).

**Comment by cr4zyc4t** [comment]
Just remove StrictMode and it will play nicely with React 18

**Comment by thednp** [comment]
> Just remove StrictMode and it will play nicely with React 18

Does this change sliminate the need to use `useSyncExternalStore`?
